src/rust/ide/src/ide.rs

```diff
@@ -0,0 +1,184 @@
+        info!(logger, "Creating a new project named `{name}`.");
+        let id = project_manager.create_project(name).await?.project_id;
+        Ok(ProjectMetadata {
+            id,
+            name        : ProjectName {name:name.clone()},
+            last_opened : None,
+        })
+    }
+
+    /// Open the named project or create a new project if it doesn't exist.
+    pub async fn open_project_or_create_new
+    ( logger : &Logger
+    , project_manager : &impl project_manager::API
+    , project_name    : &String) -> FallibleResult<controller::Project> {
+        let projects_to_list = constants::MAXIMUM_LISTABLE_PROJECTS;
+        let response     = project_manager.list_recent_projects(&projects_to_list).await?;
+        let mut projects = response.projects.iter();
+        let project      = projects.find(|project_metadata| {
+            project_metadata.name.name == *project_name
+        });
+        let project_metadata = if let Some(project) = project {
+            project.clone()
+        } else {
+            println!("Attempting to create {}", project_name);
+            Self::create_project(logger,project_manager,project_name).await?
+        };
+        let endpoints = project_manager.open_project(&project_metadata.id).await?;
+        Self::open_project(logger,endpoints.language_server_json_address,
+            endpoints.language_server_binary_address,&project_metadata.name.name).await
+    }
+
+    /// Open most recent project or create a new project if none exists.
+    pub async fn open_most_recent_project_or_create_new
+    ( logger          : &Logger
+    , project_manager : &impl project_manager::API) -> FallibleResult<controller::Project> {
+        let projects_to_list = 1;
+        let mut response     = project_manager.list_recent_projects(&projects_to_list).await?;
+        let project_metadata = if let Some(project) = response.projects.pop() {
+            project
+        } else {
+            let project_name = constants::DEFAULT_PROJECT_NAME.to_string();
+            Self::create_project(logger,project_manager,&project_name).await?
+        };
+        let endpoints = project_manager.open_project(&project_metadata.id).await?;
+        Self::open_project(logger,endpoints.language_server_json_address,
+                     endpoints.language_server_binary_address,&project_metadata.name.name).await
+    }
+
+    async fn initialize_project_manager(&mut self) -> FallibleResult<()> {
+        let transport        = self.connect_to_project_manager().await?;
+        self.project_manager = Some(Self::setup_project_manager(transport));
+        Ok(())
+    }
+
+    /// Sets up the project view, including the controller it uses.
+    pub async fn setup_project_view(&self) -> Result<ProjectView,failure::Error> {
+        let logger       = &self.logger;
+        let pm           = self.project_manager.as_ref().expect("Couldn't get Project Manager.");
```

Comment:
This error is not self explanatory. What does it mean that you couldn't "get" project manager?